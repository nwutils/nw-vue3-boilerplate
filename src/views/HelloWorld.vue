<template>
  <div class="greetings">
    <div class="logos">
      <img
        v-if="isDesktop"
        alt="NW.js logo"
        src="../assets/nw.png"
      />
      <img
        alt="Vue logo"
        src="../assets/logo.svg"
      />
    </div>

    <h1 class="green" data-test="appName">{{ APP_NAME }}</h1>
    <h2 class="green" data-test="message">{{ message }}</h2>
    <h3>
      This boilerplate uses
      <ExternalLink name="NW.js" url="nwjs.io" /> +
      <ExternalLink name="Vite" url="vitejs.dev" /> +
      <ExternalLink name="Vue 3" url="vuejs.org" /> +
      <ExternalLink name="Pinia" url="pinia.vuejs.org" /> +
      <ExternalLink name="Vue-Router" url="router.vuejs.org" /> +
      the <ExternalLink name="superior Options API" url="docs.google.com/presentation/d/1fsDhn_MptD6c-_ALgurQtvaRFkfbfukbbFGfEfckzvk/edit?usp=sharing" />.
    </h3>
    <div v-if="isDesktop" class="center mt-1">
      <p v-if="isDevMode" class="container">
        <button
          data-test="devToolsButton"
          @click="nw.Window.get().showDevTools()"
        >
          Open DevTools
        </button>
      </p>
      <p class="container">
        <button
          data-test="openAppInBrowserButton"
          @click="openAppInBrowser"
        >
          Open App in default browser
        </button>
      </p>
    </div>
    <p v-else>
      Notice that the <strong>same app conditionally renders</strong> different content
      in a normal browser than <strong>when ran as a desktop app</strong>.
    </p>
  </div>
</template>

<script>
import { APP_NAME } from '@/helpers/constants.js';

import ExternalLink from '@/components/ExternalLink.vue';

export default {
  name: 'HelloWorld',
  components: {
    ExternalLink
  },
  constants: {
    APP_NAME
  },
  methods: {
    openAppInBrowser: function () {
      if (this.isDesktop) {
        this.nw.Shell.openExternal(window.location.href);
      }
    }
  },
  computed: {
    isDevMode: function () {
      return this.isDesktop && this.process.versions['nw-flavor'] === 'sdk';
    },
    message: function () {
      if (this.isDesktop) {
        return 'NW.js & Vue Boilerplate';
      }
      return 'Vue Boilerplate';
    }
  }
};
</script>

<style scoped>
.logos {
  display: flex;
  align-items: center;
  justify-content: center;
}
.logos img {
  max-width: 160px;
  max-height: 160px;
  margin: 1rem 1rem 1rem 1rem;
}

h1 {
  font-size: 2.6rem;
  font-weight: 700;
}

h2 {
  font-size: 2.1rem;
}

h3 {
  font-size: 1.2rem;
}

.greetings h1,
.greetings h2,
.greetings h3 {
  text-align: center;
}
</style>
