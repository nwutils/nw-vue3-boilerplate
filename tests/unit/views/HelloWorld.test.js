import HelloWorld from '@/views/HelloWorld.vue';

import testHelpers from '@@/unit/testHelpers.js';

describe('HelloWorld.vue', () => {
  const setupWrapper = async () => {
    const wrapper = await testHelpers.shallowMount(HelloWorld);
    return wrapper;
  };

  describe('Desktop', () => {
    test('Render default contents', async () => {
      const wrapper = await setupWrapper();

      expect(wrapper)
        .toMatchSnapshot();
    });

    test('App name and message displays correctly', async () => {
      const wrapper = await setupWrapper();

      expect(wrapper.find('[data-test="appName"]').text())
        .toEqual('Your App Name Here');

      expect(wrapper.find('[data-test="message"]').text())
        .toEqual('NW.js & Vue Boilerplate');
    });

    test('Open dev tools', async () => {
      const wrapper = await setupWrapper();

      const button = wrapper.find('[data-test="devToolsButton"]');
      await button.trigger('click');

      expect(window.nw.Window.get().showDevTools)
        .toHaveBeenCalled();
    });

    test('App opens in default browser', async () => {
      const wrapper = await setupWrapper();

      const button = wrapper.find('[data-test="openAppInBrowserButton"]');
      await button.trigger('click');

      expect(window.nw.Shell.openExternal)
        .toHaveBeenCalledWith('http://localhost:3000/#/');
    });
  });

  describe('Web', () => {
    beforeEach(() => {
      window.webSetup();
    });

    test('Render default contents', async () => {
      const wrapper = await setupWrapper();

      expect(wrapper)
        .toMatchSnapshot();
    });

    test('Message displays correctly', async () => {
      const wrapper = await setupWrapper();

      expect(wrapper.find('[data-test="message"]').text())
        .toEqual('Vue Boilerplate');
    });
  });
});
