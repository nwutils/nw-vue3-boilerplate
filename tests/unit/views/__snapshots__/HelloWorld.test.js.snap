// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HelloWorld.vue > Desktop > Render default contents 1`] = `
<div class="greetings">
  <div class="logos">
    <img
      alt="NW.js logo"
      src="/src/assets/nw.png"
    />
    <img
      alt="Vue logo"
      src="/src/assets/logo.svg"
    />
  </div>
  <h1 class="green">
    Your App Name Here
  </h1>
  <h2 class="green">
    NW.js &amp; Vue Boilerplate
  </h2>
  <h3>
    This boilerplate uses
    <external-link-stub
      name="NW.js"
      url="nwjs.io"
    ></external-link-stub>
    +
    <external-link-stub
      name="Vite"
      url="vitejs.dev"
    ></external-link-stub>
    +
    <external-link-stub
      name="Vue 3"
      url="vuejs.org"
    ></external-link-stub>
    +
    <external-link-stub
      name="Pinia"
      url="pinia.vuejs.org"
    ></external-link-stub>
    +
    <external-link-stub
      name="Vue-Router"
      url="router.vuejs.org"
    ></external-link-stub>
    + the
    <external-link-stub
      name="superior Options API"
      url="docs.google.com/presentation/d/1fsDhn_MptD6c-_ALgurQtvaRFkfbfukbbFGfEfckzvk/edit?usp=sharing"
    ></external-link-stub>
    .
  </h3>
  <div class="
    center
    mt-1
  ">
    <p class="container">
      <button>
        Open DevTools
      </button>
    </p>
    <p class="container">
      <button>
        Open App in default browser
      </button>
    </p>
  </div>
</div>
`;

exports[`HelloWorld.vue > Web > Render default contents 1`] = `
<div class="greetings">
  <div class="logos">
    <!-- v-if -->
    <img
      alt="Vue logo"
      src="/src/assets/logo.svg"
    />
  </div>
  <h1 class="green">
    Your App Name Here
  </h1>
  <h2 class="green">
    Vue Boilerplate
  </h2>
  <h3>
    This boilerplate uses
    <external-link-stub
      name="NW.js"
      url="nwjs.io"
    ></external-link-stub>
    +
    <external-link-stub
      name="Vite"
      url="vitejs.dev"
    ></external-link-stub>
    +
    <external-link-stub
      name="Vue 3"
      url="vuejs.org"
    ></external-link-stub>
    +
    <external-link-stub
      name="Pinia"
      url="pinia.vuejs.org"
    ></external-link-stub>
    +
    <external-link-stub
      name="Vue-Router"
      url="router.vuejs.org"
    ></external-link-stub>
    + the
    <external-link-stub
      name="superior Options API"
      url="docs.google.com/presentation/d/1fsDhn_MptD6c-_ALgurQtvaRFkfbfukbbFGfEfckzvk/edit?usp=sharing"
    ></external-link-stub>
    .
  </h3>
  <p>
    Notice that the
    <strong>
      same app conditionally renders
    </strong>
    different content in a normal browser than
    <strong>
      when ran as a desktop app
    </strong>
    .
  </p>
</div>
`;
